core/src/test/java/org/apache/iceberg/avro/TestBuildAvroProjection.java

```diff
@@ -0,0 +1,244 @@
+        .prop(AvroSchemaUtil.KEY_ID_PROP, "0")
+        .prop(AvroSchemaUtil.VALUE_ID_PROP, "1")
+        .values(
+            SchemaBuilder.record("value").namespace("unit.test").fields()
+                .name("int1").prop(AvroSchemaUtil.FIELD_ID_PROP, "1").type().nullable().intType().noDefault()
+                .name("string1").prop(AvroSchemaUtil.FIELD_ID_PROP, "2").type().nullable().stringType().noDefault()
+                .endRecord());
+
+    final BuildAvroProjection testSubject = new BuildAvroProjection(icebergType, Collections.emptyMap());
+
+    final Supplier<org.apache.avro.Schema> supplier = expected::getValueType;
+
+    final org.apache.avro.Schema actual = testSubject.map(expected, supplier);
+
+    assertEquals(expected, actual);
+    assertEquals(0, Integer.valueOf(actual.getProp(AvroSchemaUtil.KEY_ID_PROP)).intValue());
+    assertEquals(1, Integer.valueOf(actual.getProp(AvroSchemaUtil.VALUE_ID_PROP)).intValue());
+  }
+
+  @Test
+  public void projectMapWithExtraFieldInValueSchema() {
+
+    final Type icebergType = Types.MapType.ofRequired(0, 1,
+        Types.StringType.get(),
+        Types.StructType.of(
+            optional(2, "int1", Types.IntegerType.get()),
+            optional(3, "string1", Types.StringType.get())
+        )
+    );
+
+    final org.apache.avro.Schema extraField = SchemaBuilder.map()
+        .prop(AvroSchemaUtil.KEY_ID_PROP, "0")
+        .prop(AvroSchemaUtil.VALUE_ID_PROP, "1")
+        .values(
+            SchemaBuilder.record("value").namespace("unit.test").fields()
+                .name("int1").prop(AvroSchemaUtil.FIELD_ID_PROP, "1").type().nullable().intType().noDefault()
+                .name("string1").prop(AvroSchemaUtil.FIELD_ID_PROP, "2").type().nullable().stringType().noDefault()
+                .name("float1").prop(AvroSchemaUtil.FIELD_ID_PROP, "3").type().nullable().floatType().noDefault()
+                .endRecord());
+
+    // once projected onto iceberg schema, the avro schema will lose the extra float field
+    final org.apache.avro.Schema expected = SchemaBuilder.map()
+        .prop(AvroSchemaUtil.KEY_ID_PROP, "0")
+        .prop(AvroSchemaUtil.VALUE_ID_PROP, "1")
+        .values(
+            SchemaBuilder.record("value").namespace("unit.test").fields()
+                .name("int1").prop(AvroSchemaUtil.FIELD_ID_PROP, "1").type().nullable().intType().noDefault()
+                .name("string1").prop(AvroSchemaUtil.FIELD_ID_PROP, "2").type().nullable().stringType().noDefault()
+                .endRecord());
+
+    final BuildAvroProjection testSubject = new BuildAvroProjection(icebergType, Collections.emptyMap());
+
+    final Supplier<org.apache.avro.Schema> supplier = expected::getValueType;
+
+    final org.apache.avro.Schema actual = testSubject.map(extraField, supplier);
+
+    assertEquals(expected, actual);
+    assertEquals(0, Integer.valueOf(actual.getProp(AvroSchemaUtil.KEY_ID_PROP)).intValue());
```

Comment:
I think most of the Iceberg unit test use asserts with a message so it's slightly more clear what fails.

Reply:
Added message for all asserts added.